### DISCUSSION TITLE
Live Preview for Drafts: Seemingly no way to fetch the version ID, can’t live preview drafts with custom endpoints

### DISCUSSION THREAD

**Discussion by robinscholz** [payloadcms/payload#15173] [Q&A]
We have built our app around custom endpoints for each collection and have also implemented live preview. This works okay for published (live) data.

In order to live preview drafts, we would need to pass a draft id to our frontend via a query in the live preview url. This query could then be forwarded to our custom endpoint and the endpoint could fetch the draft version instead of the published version. It seems that the draft id never gets passed into the arguments for creating a live preview url, correct?

I also could not find any info in the docs regarding live preview and drafts? Is this even possible even when using the built in REST API?

**Comment by jacobsfletch** [comment]
Are you implementing client-side or server-side live preview?

Typically with client-side live preview, you receive the `data` arg in your front-end via the `useLivePreview` hook. No additional data fetching is required. You simply take this data and re-render your page with it.

On the other hand, if you're implementing some custom data fetching strategy, you should be able to inject custom search params as needed using a [dynamic url](https://payloadcms.com/docs/live-preview/overview#dynamic-urls). Your frontend should be able to read these values and behave accordingly. 

Something like this (untested):

```ts
import { buildConfig } from 'payload'

const config = buildConfig({
  // ...
  admin: {
    // ...
    livePreview: {
      // ...
      url: ({ data }) => `/${data.slug}?id=${data.id}`,
    },
  },
})
```

That being said, this is not your version ID, but your document ID. Is there a reason why you'd need this at all vs just setting `draft=true` in your queries?

**Comment by robinscholz** [comment]
We are usually writing our frontend applications with Nuxt.js and expose data through custom endpoints within payload. This ensures that we are able to format data server-side exactly how we need it inside our frontend. We basically create a malleable layer between the two ends.

This sadly makes it close to impossible to update data in real-time via [window.postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage), since the structure of the updated data that is being sent by payload is totally different than the one we ingest into our frontend.

What we ended up doing is rely on autosave and a custom composable in vue:

``` ts
import { isDocumentEvent, ready } from '@payloadcms/live-preview'

export function useLivePreview() {
  const hasSentReadyMessage = ref(false)
  const hasEventListener = ref(false)

  const { query } = useRoute()

  const config = useRuntimeConfig()
  const serverURL = config.public.payload.serverURL

  async function onMessage(event: MessageEvent) {
    if (isDocumentEvent(event, serverURL)) {
      refreshNuxtData()
    }
  }

  function initialize() {
    if (!query.preview) {
      return
    }

    if (typeof window !== 'undefined' && !hasEventListener.value) {
      window.addEventListener('message', onMessage)
      hasEventListener.value = true
    }

    if (!hasSentReadyMessage.value) {
      hasSentReadyMessage.value = true

      ready({
        serverURL,
      })
    }
  }

  function destroy() {
    if (typeof window !== 'undefined') {
      window.removeEventListener('message', onMessage)
      hasEventListener.value = false
    }
  }

  onMounted(() => {
    initialize()
  })

  onUnmounted(() => {
    destroy()
  })

  watchEffect(() => {
    initialize()
  })
}
```

This essentially tells nuxt to refresh all its data fetching calls every time payload sends a message.
